Port _WKWebExtensionDeclarativeNetRequestSQLiteStore to C++

ItsJamie9494 opened the pull request:
#### 6b884b71e2361eb9bdbf586b61e9605bddc6b495
<pre>
Port _WKWebExtensionDeclarativeNetRequestSQLiteStore to C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=296080">https://bugs.webkit.org/show_bug.cgi?id=296080</a>

Reviewed by Timothy Hatcher.

Following previous work with C++ SQLite databases, this ports
the Declarative Net Request SQLite store to C++ and removes the previous Cocoa-only implementation.

* Source/WebKit/Shared/Extensions/WebExtensionSQLiteStore.cpp:
(WebKit::WebExtensionSQLiteStore::createSavepoint):
(WebKit::WebExtensionSQLiteStore::commitSavepoint):
(WebKit::WebExtensionSQLiteStore::rollbackToSavepoint):
* Source/WebKit/Shared/Extensions/WebExtensionSQLiteStore.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/Extensions/API/WebExtensionContextAPIDeclarativeNetRequest.cpp: Added.
(WebKit::WebExtensionContext::isDeclarativeNetRequestMessageAllowed):
(WebKit::WebExtensionContext::declarativeNetRequestGetEnabledRulesets):
(WebKit::WebExtensionContext::declarativeNetRequestValidateRulesetIdentifiers):
(WebKit::WebExtensionContext::declarativeNetRequestDynamicRulesStore):
(WebKit::WebExtensionContext::declarativeNetRequestSessionRulesStore):
(WebKit::WebExtensionContext::updateDeclarativeNetRequestRulesInStorage):
(WebKit::WebExtensionContext::declarativeNetRequestGetDynamicRules):
(WebKit::WebExtensionContext::declarativeNetRequestUpdateDynamicRules):
(WebKit::WebExtensionContext::declarativeNetRequestGetSessionRules):
(WebKit::WebExtensionContext::declarativeNetRequestUpdateSessionRules):
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm:
(WebKit::WebExtensionContext::isDeclarativeNetRequestMessageAllowed): Deleted.
(WebKit::WebExtensionContext::declarativeNetRequestGetEnabledRulesets): Deleted.
(WebKit::WebExtensionContext::declarativeNetRequestValidateRulesetIdentifiers): Deleted.
(WebKit::WebExtensionContext::declarativeNetRequestDynamicRulesStore): Deleted.
(WebKit::WebExtensionContext::declarativeNetRequestSessionRulesStore): Deleted.
(WebKit::WebExtensionContext::updateDeclarativeNetRequestRulesInStorage): Deleted.
(WebKit::WebExtensionContext::declarativeNetRequestGetDynamicRules): Deleted.
(WebKit::WebExtensionContext::declarativeNetRequestUpdateDynamicRules): Deleted.
(WebKit::WebExtensionContext::declarativeNetRequestGetSessionRules): Deleted.
(WebKit::WebExtensionContext::declarativeNetRequestUpdateSessionRules): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::loadDeclarativeNetRequestRules):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDeclarativeNetRequestSQLiteStoreCocoa.mm: Added.
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::migrateData):
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestSQLiteStore.h: Removed.
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestSQLiteStore.mm: Removed.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionDeclarativeNetRequestSQLiteStore.cpp:
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::WebExtensionDeclarativeNetRequestSQLiteStore):
(WebKit::ruleIdMapToString):
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::updateRulesByRemovingIDs):
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::addRules):
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::deleteRules):
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::getRulesWithRuleIDs):
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::getRulesWithRuleIDsInternal):
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::getKeysAndValuesFromRowIterator):
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::insertRule):
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::currentDatabaseSchemaVersion):
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::databaseURL):
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::createFreshDatabaseSchema):
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::migrateToCurrentSchemaVersionIfNeeded):
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::resetDatabaseSchema):
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::isDatabaseEmpty):
* Source/WebKit/UIProcess/Extensions/WebExtensionDeclarativeNetRequestSQLiteStore.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionStorageSQLiteStore.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/DeclarativeNetRequestRules.db: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/DeclarativeNetRequestRules.db-shm: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/DeclarativeNetRequestRules.db-wal: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIDeclarativeNetRequest.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, MigrateDeclarativeNetRequestDataToNewFormat)):

Canonical link: <a href="https://commits.webkit.org/300056@main">https://commits.webkit.org/300056@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c11aa08a3475e6fbf3c11c1ac2e149b02b46e5b4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121166 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/40862 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31520 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127590 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73247 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/123042 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41564 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49441 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92047 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/61239 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124118 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33188 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108607 "Found 1 new API test failure: TestIPC.StreamServerConnectionTests/StreamServerDidReceiveInvalidMessageTest.AsyncNotStreamEncodable/ValidationError (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72723 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32214 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26710 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71179 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102690 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26891 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130438 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48093 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36557 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100645 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48461 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104776 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100549 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45952 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24007 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/44787 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19224 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/47951 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53664 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47422 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50769 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49106 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionDeclarativeNetRequestSQLiteStore.cpp:
```diff
@@ -0,0 +1,354 @@
+    assertIsCurrent(queue());
+
+    auto currentDatabaseSchemaVersion = databaseSchemaVersion();
+    if (currentDatabaseSchemaVersion == 1 && currentDeclarativeNetRequestSchemaVersion == 2) {
```
I think we should drop the `currentDeclarativeNetRequestSchemaVersion == 2`. What happens when the current versions is 3 and someone is on version 1?

```suggestion
    if (currentDatabaseSchemaVersion == 1) {
```

xeenon reviewed (changes requested):
Looking good, just one question about the migration version check.

ItsJamie9494 committed:
Port _WKWebExtensionDeclarativeNetRequestSQLiteStore to C++
https://bugs.webkit.org/show_bug.cgi?id=296080

Reviewed by Timothy Hatcher.

Following previous work with C++ SQLite databases, this ports
the Declarative Net Request SQLite store to C++ and removes the previous Cocoa-only implementation.

* Source/WebKit/Shared/Extensions/WebExtensionSQLiteStore.cpp:
(WebKit::WebExtensionSQLiteStore::createSavepoint):
(WebKit::WebExtensionSQLiteStore::commitSavepoint):
(WebKit::WebExtensionSQLiteStore::rollbackToSavepoint):
* Source/WebKit/Shared/Extensions/WebExtensionSQLiteStore.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/Extensions/API/WebExtensionContextAPIDeclarativeNetRequest.cpp: Added.
(WebKit::WebExtensionContext::isDeclarativeNetRequestMessageAllowed):
(WebKit::WebExtensionContext::declarativeNetRequestGetEnabledRulesets):
(WebKit::WebExtensionContext::declarativeNetRequestValidateRulesetIdentifiers):
(WebKit::WebExtensionContext::declarativeNetRequestDynamicRulesStore):
(WebKit::WebExtensionContext::declarativeNetRequestSessionRulesStore):
(WebKit::WebExtensionContext::updateDeclarativeNetRequestRulesInStorage):
(WebKit::WebExtensionContext::declarativeNetRequestGetDynamicRules):
(WebKit::WebExtensionContext::declarativeNetRequestUpdateDynamicRules):
(WebKit::WebExtensionContext::declarativeNetRequestGetSessionRules):
(WebKit::WebExtensionContext::declarativeNetRequestUpdateSessionRules):
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm:
(WebKit::WebExtensionContext::isDeclarativeNetRequestMessageAllowed): Deleted.
(WebKit::WebExtensionContext::declarativeNetRequestGetEnabledRulesets): Deleted.
(WebKit::WebExtensionContext::declarativeNetRequestValidateRulesetIdentifiers): Deleted.
(WebKit::WebExtensionContext::declarativeNetRequestDynamicRulesStore): Deleted.
(WebKit::WebExtensionContext::declarativeNetRequestSessionRulesStore): Deleted.
(WebKit::WebExtensionContext::updateDeclarativeNetRequestRulesInStorage): Deleted.
(WebKit::WebExtensionContext::declarativeNetRequestGetDynamicRules): Deleted.
(WebKit::WebExtensionContext::declarativeNetRequestUpdateDynamicRules): Deleted.
(WebKit::WebExtensionContext::declarativeNetRequestGetSessionRules): Deleted.
(WebKit::WebExtensionContext::declarativeNetRequestUpdateSessionRules): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::loadDeclarativeNetRequestRules):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDeclarativeNetRequestSQLiteStoreCocoa.mm: Added.
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::migrateData):
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestSQLiteStore.h: Removed.
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestSQLiteStore.mm: Removed.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionDeclarativeNetRequestSQLiteStore.cpp:
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::WebExtensionDeclarativeNetRequestSQLiteStore):
(WebKit::ruleIdMapToString):
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::updateRulesByRemovingIDs):
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::addRules):
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::deleteRules):
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::getRulesWithRuleIDs):
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::getRulesWithRuleIDsInternal):
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::getKeysAndValuesFromRowIterator):
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::insertRule):
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::currentDatabaseSchemaVersion):
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::databaseURL):
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::createFreshDatabaseSchema):
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::migrateToCurrentSchemaVersionIfNeeded):
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::resetDatabaseSchema):
(WebKit::WebExtensionDeclarativeNetRequestSQLiteStore::isDatabaseEmpty):
* Source/WebKit/UIProcess/Extensions/WebExtensionDeclarativeNetRequestSQLiteStore.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionStorageSQLiteStore.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/DeclarativeNetRequestRules.db: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/DeclarativeNetRequestRules.db-shm: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/DeclarativeNetRequestRules.db-wal: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIDeclarativeNetRequest.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, MigrateDeclarativeNetRequestDataToNewFormat)):

Canonical link: https://commits.webkit.org/300056@main
